tests/asyncio/retry/test_retry_streaming_async.py

```diff
@@ -0,0 +1,541 @@
+        retry_ = retry_streaming_async.AsyncStreamingRetry()
+        decorated = retry_(self._generator_mock)
+        exception_list = []
+        generator = await decorated(10, exceptions_seen=exception_list)
+        await generator.aclose()
+
+        with pytest.raises(StopAsyncIteration):
+            # calling next on closed generator should raise error
+            await generator.__anext__()
+
+    @mock.patch("asyncio.sleep", autospec=True)
+    @pytest.mark.asyncio
+    async def test___call___with_generator_throw(self, sleep):
+        """
+        Throw should be passed through retry into target generator
+        """
+
+        # The generator should not retry when it encounters a non-retryable error
+        retry_ = retry_streaming_async.AsyncStreamingRetry(
+            predicate=retry_async.if_exception_type(ValueError),
+        )
+        decorated = retry_(self._generator_mock)
+        exception_list = []
+        generator = await decorated(10, exceptions_seen=exception_list)
+        for i in range(2):
+            await generator.__anext__()
+        with pytest.raises(BufferError):
+            await generator.athrow(BufferError("test"))
+        assert isinstance(exception_list[0], BufferError)
+        with pytest.raises(StopAsyncIteration):
+            # calling next on closed generator should raise error
+            await generator.__anext__()
+
+        # In contrast, the generator should retry if we throw a retryable exception
+        exception_list = []
+        generator = await decorated(10, exceptions_seen=exception_list)
+        for i in range(2):
+            await generator.__anext__()
+        throw_val = await generator.athrow(ValueError("test"))
+        assert throw_val == 0
+        assert isinstance(exception_list[0], ValueError)
+        # calling next on generator should not raise error, because it was retried
+        assert await generator.__anext__() == 1
+
+    @pytest.mark.parametrize("awaitable_wrapped", [True, False])
+    @mock.patch("asyncio.sleep", autospec=True)
+    @pytest.mark.asyncio
+    async def test___call___with_iterable_send(self, sleep, awaitable_wrapped):
+        """
+        Send should work like next if the wrapped iterable does not support it
+        """
+        retry_ = retry_streaming_async.AsyncStreamingRetry()
+
+        def iterable_fn():
+            class CustomIterable:
+                def __init__(self):
+                    self.i = -1
+
+                def __aiter__(self):
+                    return self
+
+                async def __anext__(self):
+                    self.i += 1
+                    return self.i
+
+            return CustomIterable()
+
+        if awaitable_wrapped:
+
+            async def wrapper():
+                return iterable_fn()
+
+            decorated = retry_(wrapper)
+        else:
+            decorated = retry_(iterable_fn)
+
+        retryable = await decorated()
+        result = await retryable.__anext__()
```

Comment:
should this be `asend`? (Honest question; I'm still murky on this)

Reply:
You can't call send until it's been "started" (i.e., next has been called once). I'll add a comment around this